x/staking/handler.go

```diff
@@ -245,8 +278,42 @@ func handleMsgUndelegate(ctx sdk.Context, msg types.MsgUndelegate, k keeper.Keep
 }
 
 func handleMsgBeginRedelegate(ctx sdk.Context, msg types.MsgBeginRedelegate, k keeper.Keeper) sdk.Result {
-	completionTime, err := k.BeginRedelegation(ctx, msg.DelegatorAddress, msg.ValidatorSrcAddress,
-		msg.ValidatorDstAddress, msg.SharesAmount)
+	validator, found := k.GetValidator(ctx, msg.ValidatorSrcAddress)
+	if !found {
+		return ErrNoValidatorFound(k.Codespace()).Result()
+	}
+
+	del, found := k.GetDelegation(ctx, msg.DelegatorAddress, msg.ValidatorSrcAddress)
+	if !found {
+		return ErrNoDelegation(k.Codespace()).Result()
+	}
+
+	shares, err := validator.SharesFromTokens(msg.Amount.Amount)
```

Comment:
Can we separate this logic, it seems identical to that in `handleMsgUndelegate` (about the next 10 lines)

Reply:
Updated and move logic to `Keeper#ValidateUnbondAmount`.